nav2_costmap_2d/src/costmap_cleaner.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <memory>
+
+#include "nav2_costmap_2d/costmap_cleaner.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+namespace nav2_costmap_2d
+{
+
+using std::vector;
+using std::string;
+using std::shared_ptr;
+using std::any_of;
+using CleanCostmap = nav2_msgs::srv::CleanCostmap;
+
+CostmapCleaner::CostmapCleaner(rclcpp::Node::SharedPtr & node, Costmap2DROS * costmap)
+: node_(node), costmap_(costmap)
+{
+  node_->get_parameter_or_set("cleanable_layers", cleanable_layers_, {"obstacle_layer"});
```

Comment:
I'd rather the clear service just called all the layers and told them to clear. Leave it up to the layer to decide whether it can be cleared or not.

Looks like costmap_ros already has a `resetLayers()` function. Why not just call that?

Reply:
Clearing/resetting with those methods clears all the layers entirely.